src/Entity/SnapshotManager.php

```diff
@@ -236,60 +236,28 @@ public function cleanup(PageInterface $page, $keep)
             throw new \RuntimeException(sprintf('Please provide an integer value, %s given', \gettype($keep)));
         }
 
-        $tableName = $this->getTableName();
-        $platform = $this->getConnection()->getDatabasePlatform()->getName();
-
-        if ('mysql' === $platform) {
-            return $this->getConnection()->exec(sprintf(
-                'DELETE FROM %s
-                WHERE
-                    page_id = %d
-                    AND id NOT IN (
-                        SELECT id
-                        FROM (
-                            SELECT id, publication_date_end
-                            FROM %s
-                            WHERE
-                                page_id = %d
-                            ORDER BY
-                                publication_date_end IS NULL DESC,
-                                publication_date_end DESC
-                            LIMIT %d
-                        ) AS table_alias
-                )',
-                $tableName,
-                $page->getId(),
-                $tableName,
-                $page->getId(),
-                $keep
-            ));
-        }
+        $em = $this->getEntityManager();
 
-        if ('oracle' === $platform) {
-            return $this->getConnection()->exec(sprintf(
-                'DELETE FROM %s
-                WHERE
-                    page_id = %d
-                    AND id NOT IN (
-                        SELECT id
-                        FROM (
-                            SELECT id, publication_date_end
-                            FROM %s
-                            WHERE
-                                page_id = %d
-                                AND rownum <= %d
-                            ORDER BY publication_date_end DESC
-                        ) table_alias
-                )',
-                $tableName,
-                $page->getId(),
-                $tableName,
-                $page->getId(),
-                $keep
-            ));
-        }
+        $snapShots = $em->createQueryBuilder()
+            ->select('s.id')
+            ->from($this->class, 's')
+            ->where('s.page = :page')
+            ->orderBy('s.publicationDateEnd', 'DESC')
+            ->setParameter('page', $page)
+            ->setMaxResults($keep)
+            ->getQuery()
+            ->getArrayResult();
+
+        $query = $em->createQueryBuilder()
+            ->delete()
+            ->from($this->class, 's')
+            ->where('s.page = :page')
+            ->andWhere('s.id NOT IN (:pageId)')
+            ->setParameter('page', $page)
+            ->setParameter('pageId', array_column($snapShots, 'id'))
+            ->getQuery();
```

Comment:
Why use a query builder here over plain DQL? There are no conditional clauses…

Reply:
Probably just out of habit to use a query builder, it's a personal preference to use a query builder over DQL.
But if plain DQL is the preferred way then I'll happily change it

Reply:
Please do. Plain DQL is not the preferred way, because you should pick different APIs for different use cases. Here is a table to make your decision: http://talks.nekland.fr/Doctrine/#/32 (:fr:)